File: tests/typesafety/python/api_parse_misuse.py
# expect: fails
# title: A bare dictionary where an Order is required, rejected by the checker
# misuse-of: api_parse
from pydantic import BaseModel, ConfigDict


class Order(BaseModel):
    model_config = ConfigDict(strict=True)

    id: str
    quantity: int
    gift_note: str | None


def price_cents(order: Order) -> int:
    return order.quantity * 250


total = price_cents({"id": "o1", "quantity": "2", "gift_note": None})

File: tests/typesafety/typescript/api_parse_misuse.ts
// expect: fails

import { z } from "zod";

const Order = z.strictObject({
  id: z.string(),
  quantity: z.number().int(),
  giftNote: z.string().nullable(),
});

type Order = z.infer<typeof Order>;

function priceCents(order: Order): number {
  return order.quantity * 250;
}

export const total = priceCents({ id: "o1", quantity: "2", giftNote: null });
